specs/casper_sharding_v2.1.md

```diff
@@ -581,27 +581,43 @@ Finally:
 * Let `next_start_shard = (shard_and_committee_for_slots[-1][-1].shard_id + 1) % SHARD_COUNT`
 * Set `shard_and_committee_for_slots[CYCLE_LENGTH:] = get_new_shuffling(block.ancestor_hashes[0], validators, next_start_shard)`
 
--------
-
-Note: this is ~80% complete. The main sections that are missing are:
-
-* Logic for the formats of shard chains, who proposes shard blocks, etc. (in an initial release, if desired we could make crosslinks just be Merkle roots of blobs of data; in any case, one can philosophically view the whole point of the shard chains as being a coordination device for choosing what blobs of data to propose as crosslinks)
-* Logic for inducting queued validators from the PoW chain
-* Penalties for signing or attesting to non-canonical-chain blocks (update: may not be necessary, see https://ethresear.ch/t/attestation-committee-based-full-pos-chains/2259)
-* Per-validator proofs of custody, and associated slashing conditions
-* Versioning and upgrades
-
-Slashing conditions may include:
-
-
-    Casper FFG slot equivocation [done]
-    Casper FFG surround [done]
-    Beacon chain proposal equivocation [done]
-    Shard chain proposal equivocation
-    Proof of custody secret leak
-    Proof of custody wrong custody bit
-    Proof of custody no secret reveal
-    RANDAO leak
+### TODO
+
+Note: This spec is ~60% complete.
+
+* [ ] Specify how `crystallized_state_root` and `active_state_root` are constructed
+* [ ] Specify the shard chain blocks, blobs, proposers, etc.
+* [ ] Specify the rules for forced deregistrations
+* [ ] Add a double-batched Merkle accumulator for historical beacon chain blocks
+* [ ] Specify the various assumptions (global clock, validator honesty, validator liveness, etc.)
+* [ ] Specify in Vyper the registration contract on the PoW chain
+* [ ] Specify the RANDAO logic, including the hardening against orphaned reveals
+* [ ] Add per-validator proofs of custody, including slashing conditions
+* [ ] Use a separate hash function for the proof of possession
+* [ ] Add an appendix for BLS12-381
+* [ ] Add an appendix on gossip networks and the offchain signature aggregation logic
+* [ ] Comprehensively and clearly define all the terms in the glossary
+* [ ] Clearly document the various edge cases, e.g. with committee sizing
+* [ ] Rework the document for readability
+* [ ] Undergo peer review, security audits and formal verification
+
+**Possible changes and additions**
+
+* [ ] Deprecate Wei and use 64-bit balances, consistent with a 64-bit EVM2.0
+* [ ] Allow for deposits larger than 32 ETH, as well as deposit top ups
+* [ ] Add penalties for a deposit below 32 ETH (or some other threshold)
+* [ ] Add a `SpecialObject` to change the `withdrawal_shard_id`, `withdrawal_address` or `randao_commitment`
+* [ ] Replace Blake with a STARK-friendly hash function
+* [ ] Replace the IMD fork choice rule with LMD
+* [ ] Merge `crystallized_state_root` and `active_state_root` into a single root
```

Comment:
Two roots are useful at least in construction. crystallized_state is large and changes slowly. active_state is small and changes a lot

If we merge the roots it should be `state_root = hash(crystallized_state_root + active_state_root)` so that the crystallized portion can stay pre-hashed.

Reply:
You can have a single root and still maintain the infrequent updates to the crystallized state (and the "pre-hash"). It's a contrived example, but you can simply Merklelise `root := hash(crystallized_state_root || active_state_root)`.

Reply:
right